### PR TITLE
Fix ExtractDestination() for ColdStaking

### PR DESCRIPTION
**Pull request by aguycalled**
What to test:

* Cold Staking works correctly with this patch

### CODE REVIEW

**Comment by mxaddict** [comment]
Building with Ubuntu 20.04

**Comment by mxaddict** [comment]
Testing the cold staking now:

image


Will let this run for a while, will approve once I have had the cold staking running for about an hours

**Comment by aguycalled** [comment]
Good! It should also get tested that spendable and stakable balances are correctly shown, my tests were successful on that regard.

**Comment by mxaddict** [comment]
> Good! It should also get tested that spendable and stakable balances are correctly shown, my tests were successful on that regard.

So far tha balances are correct on the UI as well, but will monitor it as it keeps staking.

**Comment by mxaddict** [comment]
@aguycalled "Pending" balance is only for when a TX adds balance but is not yet staked in a block correct?

**Comment by aguycalled** [comment]
yes

**Comment by mxaddict** [comment]
I've setup 2 cold staking nodes and 1 spending node, seems to be working nicely so far, will let these run for a while longer:

image

**Review by mxaddict — APPROVED** [review]
I think I'm happy with the testing on this PR, seems all is working with regards to cold staking, also testing some hot wallet staking and was still working as expected.

File: src/script/script.cpp [hunk 1]
```diff
@@ -231,6 +231,71 @@ bool CScript::IsColdStakingv2() const
...
+            (*this)[28] == OP_DUP &&
+            (*this)[29] == OP_HASH160 &&
+            (*this)[30] == 0x14 &&
+            (*this)[51] == OP_EQUALVERIFY &&
+            (*this)[52] == OP_CHECKSIG &&
+            (*this)[53] == OP_ENDIF);
+};
+
+bool CScript::HasColdStakingv2P2PKHStaking() const {
+    return (this->size() > 50 &&
```
Comment (RealKeyboardWarrior) [inline src/script/script.cpp:265]: must be ==

File: src/script/script.cpp [hunk 2]
```diff
@@ -231,6 +231,71 @@ bool CScript::IsColdStakingv2() const
             (*this)[23] == OP_IF);
 }
 
+bool CScript::HasColdStakingP2PKHStaking() const {
+    return (this->size() > 27 &&
```
Comment (RealKeyboardWarrior) [inline src/script/script.cpp:235]: must be ==

**Review by RealKeyboardWarrior — CHANGES_REQUESTED** [review]

File: src/script/script.cpp [hunk 3]
```diff
@@ -231,6 +231,71 @@ bool CScript::IsColdStakingv2() const
...
+            (*this)[0] == OP_COINSTAKE &&
+            (*this)[1] == OP_IF &&
+            (*this)[2] == OP_DUP &&
+            (*this)[3] == OP_HASH160 &&
+            (*this)[4] == 0x14 &&
+            (*this)[25] == OP_EQUALVERIFY &&
+            (*this)[26] == OP_CHECKSIG &&
+            (*this)[27] == OP_ELSE);
```
Comment (RealKeyboardWarrior) [inline src/script/script.cpp:243]: why is it not validated beyond this point?

File: src/script/script.cpp [hunk 4]
```diff
@@ -231,6 +231,71 @@ bool CScript::IsColdStakingv2() const
...
+            (*this)[0] == 0x14 &&
+            (*this)[21] == OP_DROP &&
+            (*this)[22] == OP_COINSTAKE &&
+            (*this)[23] == OP_IF &&
+            (*this)[24] == OP_DUP &&
+            (*this)[25] == OP_HASH160 &&
+            (*this)[26] == 0x14 &&
+            (*this)[47] == OP_EQUALVERIFY &&
+            (*this)[48] == OP_CHECKSIG &&
+            (*this)[49] == OP_ELSE);
```
Comment (RealKeyboardWarrior) [inline src/script/script.cpp:275]: ditto why is it not validated beyond this point?

**Review by RealKeyboardWarrior — CHANGES_REQUESTED** [review]

File: src/script/script.cpp (hunk 1 above)
Comment (RealKeyboardWarrior) [reply src/script/script.cpp:265]: nvm - it's just to get the coldstake key

File: src/script/script.cpp (hunk 2 above)
Comment (RealKeyboardWarrior) [reply src/script/script.cpp:235]: nvm - it's just to get the coldstake key

File: src/script/script.cpp (hunk 3 above)
Comment (RealKeyboardWarrior) [reply src/script/script.cpp:243]: nvm - it's just to get the coldstake key

File: src/script/script.cpp (hunk 4 above)
Comment (RealKeyboardWarrior) [reply src/script/script.cpp:275]: nvm - it's just to get the coldstake key

**Comment by chasingkirkjufell** [comment]
building with ubuntu 20.04 as well, will report any findings

**Comment by chasingkirkjufell** [comment]
when i try to send coins to the cold staking address the balance isn't correct.
steps:
create 2 wallets, create a coldstaking address v1 with wallet 1 benign spending and wallet 2 being staking
disable staking for both wallets to have better control
send multiple transaction from wallet 1 to the cold staking address
generate a block
image

**Comment by chasingkirkjufell** [comment]
so it looks like the staking wallet is showing the correct balance since actually there are only 2 groupings that went through? history shows all 5 though. window on the left is from spending wallet while on the right is the staking wallet
image

**Comment by mxaddict** [comment]
@chasingkirkjufell I did not run into any balance issues with my testing, can you give me steps to replicate?

**Comment by aguycalled** [comment]
image
 
I haven't been able to replicate it. Most likely case is that the first transactions where spent in the following (in case they were sent to a cold staking address with a spending address of that node). Will wait to hear more details to see if we can replicate/confirm.

**Comment by chasingkirkjufell** [comment]
> image
> 
> I haven't been able to replicate it. Most likely case is that the first transactions were spent in the following (in case they were sent to a cold staking address with a spending address of that node). Will wait to hear more details to see if we can replicate/confirm.

I noticed that this issue is not related to this PR so I am going to approve this PR for now. Lke your said, it may be because of the outputs were spent again but the transaction history is confusing. For what it's worth, here are a bit more detailed steps.

1. creating 2 wallets and disable staking
2. connect 2nd wallet to first
3. generate 300 blocks on 1st wallet
4. create a coldstaking address v1 with wallet 1 spending and wallet 2 staking
5. send 5 txs to coldstaking address
6. generate more blocks
7. the tab will show all 5 txes but only 2 are actually valid. repair wallet or zapwallettxes does not fix it

**Review by chasingkirkjufell — APPROVED** [review]

**Pull request merged**